### PR TITLE
sql: stop creating superfluous indexes after primary key change

### LINKED ISSUE #64838: Unexpected secondary index created when using ALTER TABLE to hash shard the primary key for the first time
**Describe the problem**

When updating a non-hash-sharded primary key to a hash sharded one, a secondary index is created on the old primary key (in addition to the correct behavior of hash sharding the index).

This is a problem because anyone hash sharding a table is trying to avoid the problem of hotspotting, but by creating this secondary index along with the update to the primary key, CockroachDB is persisting the issue and a savvy user would always want to drop the new secondary index. This is an edge case for ALTER PRIMARY KEY where the typical desired behavior ([ALTER PRIMARY KEY docs](https://www.cockroachlabs.com/docs/v20.2/alter-primary-key.html)) is unexpected.

**To Reproduce**

1. Create a table without hash sharding it:
```
CREATE TABLE events (
     product_id INT8 NOT NULL,
     owner UUID NOT NULL,
     serial_number VARCHAR NOT NULL,
     event_id UUID NULL,
     ts TIMESTAMP NULL,
     data JSONB NULL,
     PRIMARY KEY (product_id ASC, owner ASC, serial_number ASC)
 );
```

2. Update the table to hash shard it:
```
SET experimental_enable_hash_sharded_indexes TO true;  -- may be necessary

ALTER TABLE events
    ALTER PRIMARY KEY USING COLUMNS (product_id, owner, serial_number)
    USING HASH WITH BUCKET_COUNT = 5;
```

3. Perform a SHOW INDEXES command:
```
> SHOW INDEXES FROM events;
  table_name |                index_name                 | non_unique | seq_in_index |                     column_name                      | direction | storing | implicit
-------------+-------------------------------------------+------------+--------------+------------------------------------------------------+-----------+---------+-----------
  events     | primary                                   |   false    |            1 | crdb_internal_owner_product_id_serial_number_shard_5 | ASC       |  false  |  false
  events     | primary                                   |   false    |            2 | product_id                                           | ASC       |  false  |  false
  events     | primary                                   |   false    |            3 | owner                                                | ASC       |  false  |  false
  events     | primary                                   |   false    |            4 | serial_number                                        | ASC       |  false  |  false
  events     | events_product_id_owner_serial_number_key |   false    |            1 | product_id                                           | ASC       |  false  |  false
  events     | events_product_id_owner_serial_number_key |   false    |            2 | owner                                                | ASC       |  false  |  false
  events     | events_product_id_owner_serial_number_key |   false    |            3 | serial_number                                        | ASC       |  false  |  false
  events     | events_product_id_owner_serial_number_key |   false    |            4 | crdb_internal_owner_product_id_serial_number_shard_5 | ASC       |  false  |   true
```

**Expected behavior**
No secondary indexes are on the table.

**Actual behavior**
A secondary index is created, keyed on the old primary key, `(product_id, owner, serial_number)`

Note that the primary behavior, the update to the primary key, worked correctly & as expected. It's only the creation of the secondary index that's unexpected.

**Additional data / screenshots**
N/A

**Environment:**
 - CockroachDB version: 20.2.8 server, 20.2.5 for SQL client
 - Server OS: CockroachCloud
 - Client app: Cockroach sql v.20.2.5

**Additional context**
Related to [this issue](https://github.com/cockroachdb/cockroach/issues/45889).

Like that issue, this is an edge case where the typical desired behavior for ALTER TABLE (the creation of the secondary index on the old keys) isn't desired.

Epic CRDB-7363

### PR DESCRIPTION
**Pull request by ajwerner**
The logic prior to this change would maintain the old primary index as a new
unique constraint even in cases where the key columns and directions remain
the same between the new and old index. This meaningfully happens when moving
from or to hash sharding from not or from or two interleaving from not.
    
Fixes #64838.

Release note (bug fix): Fixes a bug where a superfluous unique constraint would
be added to a table during a primary key change when the primary key change
specifies a new primary key constraint that involves the same columns in the
same directions.

### CODE REVIEW

**Review by postamar — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r=postamar

File: pkg/sql/logictest/testdata/logic_test/alter_primary_key [hunk 1]
```diff
@@ -1425,3 +1456,66 @@ table_with_virtual_cols  CREATE TABLE public.table_with_virtual_cols (
...
+# Test that we do not create new indexes for the old primary key when going
+# from sharded to non-sharded and back.
+subtest toggle_sharded_no_new_index
+
+statement ok
+DROP TABLE IF EXISTS t;
+CREATE TABLE t (i INT PRIMARY KEY)
+
+query TTT
+SELECT index_name,column_name,direction FROM [SHOW INDEXES FROM t]
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/alter_primary_key:1469]: super nit: `index_name, column_name, direction`

File: pkg/sql/logictest/testdata/logic_test/alter_primary_key [hunk 2]
```diff
@@ -1425,3 +1456,66 @@ table_with_virtual_cols  CREATE TABLE public.table_with_virtual_cols (
...
+ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (i) USING HASH WITH BUCKET_COUNT = 2
+
+query TTT
+SELECT index_name,column_name,direction FROM [SHOW INDEXES FROM t]
+----
+primary  crdb_internal_i_shard_2  ASC
+primary  i                        ASC
+
+statement ok
+ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (i);
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/alter_primary_key:1483]: nit: trailing `;`

File: pkg/sql/alter_primary_key.go [hunk 3]
```diff
@@ -627,30 +627,43 @@ func (p *planner) shouldCreateIndexes(
...
-	}
-	if desc.IsPrimaryIndexDefaultRowID() {
-		return false
+
+	columnIDsAndDirsWithoutSharded := func(idx *descpb.IndexDescriptor) (
+		columnIDs descpb.ColumnIDs,
+		columnDirs []descpb.IndexDescriptor_Direction,
+	) {
+		for i, colName := range idx.KeyColumnNames {
+			if colName != idx.Sharded.Name {
```
Comment (otan) [inline pkg/sql/alter_primary_key.go:644]: maybe `columnIDs = columnsIDs[idx. ExplicitColumnStartIdx():]`, so it takes care of `crdb_region` as well?

**Comment by ajwerner** [comment]
@otan sorry I got hasty. I'll send a pr for the little stuff. Can you help me understand the scenario where omitting the region column will help? I thought we already skip if the locality config changes.

**Comment by otan** [comment]
> Can you help me understand the scenario where omitting the region column will help? I thought we already skip if the locality config changes.

oh! yeah i guess it's covered, just thought it would've been clearer. also thought this would affect the ALTER PRIMARY KEY case for a REGIONAL BY ROW table, but I guess you can't ALTER PRIMARY KEY to the same set of columns which would be the only reason it popped up.

> sorry I got hasty. I'll send a pr for the little stuff.

no biggie!

**Pull request merged**